File: Company 2 : Amazon/Day 2/5. Phone directory/code.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class GFG{
    public static void main(String args[])throws IOException
    {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(in.readLine());
        while(t-- > 0){
            int n = Integer.parseInt(in.readLine());
            String contact[] = in.readLine().trim().split("\\s+");
            String s = in.readLine();
            
            Solution ob = new Solution();
            ArrayList<ArrayList<String>> ans = ob.displayContacts(n, contact, s);
            for(int i = 0;i < ans.size();i++){
                for(int j = 0;j < ans.get(i).size();j++)
                    System.out.print(ans.get(i).get(j) + " ");
                System.out.println();
            }
        
System.out.println("~");
}
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution{
    static ArrayList<ArrayList<String>> displayContacts(int n, 
                                        String contact[], String s)
    {
        // code here
        ArrayList<ArrayList<String>> result = new ArrayList<>();
        Trie trie = new Trie();
        for(String cont : contact){
            trie.insert(cont);
        }
        
        for(int i = 1; i <= s.length(); i++){
            ArrayList<String> list = trie.getWords(s.substring(0,i));
            Collections.sort(list); // Sorting the list
            list = new ArrayList<>(new LinkedHashSet<>(list)); // Removing duplicates while maintaining orders
            if(list.isEmpty()){
                list.add("0");
            }
            result.add(list);
        }
        return result;
        
    }
    
    static class TrieNode{
        TrieNode arr[] = new TrieNode[26];
        boolean isEnd;
        ArrayList<String> words;
        TrieNode(){
            words = new ArrayList<>();
        }
    }
    
    static class Trie{
        TrieNode root;
        
        Trie(){
            root = new TrieNode();
        }
        
        void insert(String s){
            TrieNode curr = root;
            for(int i=0;i<s.length();i++){
                char ch = s.charAt(i);
                if(curr.arr[ch-'a']==null){
                    curr.arr[ch-'a'] = new TrieNode();
                }
                curr = curr.arr[ch-'a'];
                curr.words.add(s);
            }
            curr.isEnd = true;
        }
        
        ArrayList<String> getWords(String prefix){
            TrieNode curr = root;
            for(int i=0;i<prefix.length();i++){
                char ch = prefix.charAt(i);
                if(curr.arr[ch-'a']==null) return new ArrayList<>();
                curr = curr.arr[ch-'a'];
            }
            return curr.words;
        }
    }
}

File: Company 2 : Amazon/Day 2/5. Phone directory/code.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_set>
#include <sstream>

using namespace std;

class TrieNode {
public:
    TrieNode* arr[26] = {nullptr};
    bool isEnd = false;
    vector<string> words;

    TrieNode() {}
};

class Trie {
public:
    TrieNode* root;

    Trie() {
        root = new TrieNode();
    }

    void insert(const string& s) {
        TrieNode* curr = root;
        for (char ch : s) {
            if (curr->arr[ch - 'a'] == nullptr) {
                curr->arr[ch - 'a'] = new TrieNode();
            }
            curr = curr->arr[ch - 'a'];
            curr->words.push_back(s);
        }
        curr->isEnd = true;
    }

    vector<string> getWords(const string& prefix) {
        TrieNode* curr = root;
        for (char ch : prefix) {
            if (curr->arr[ch - 'a'] == nullptr) return {};
            curr = curr->arr[ch - 'a'];
        }
        return curr->words;
    }
};

class Solution {
public:
    static vector<vector<string>> displayContacts(int n, string contact[], string s) {
        vector<vector<string>> result;
        Trie trie;
        for (int i = 0; i < n; i++) {
            trie.insert(contact[i]);
        }

        for (int i = 1; i <= s.length(); i++) {
            vector<string> list = trie.getWords(s.substr(0, i));
            sort(list.begin(), list.end());
            unordered_set<string> uniqueWords(list.begin(), list.end());
            list.assign(uniqueWords.begin(), uniqueWords.end());
            if (list.empty()) {
                list.push_back("0");
            }
            result.push_back(list);
        }
        return result;
    }
};

int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t-- > 0) {
        int n;
        cin >> n;
        cin.ignore();
        string contact[n];
        for (int i = 0; i < n; i++) {
            getline(cin, contact[i]);
        }
        string s;
        getline(cin, s);

        Solution ob;
        vector<vector<string>> ans = ob.displayContacts(n, contact, s);
        for (const auto& list : ans) {
            for (const auto& word : list) {
                cout << word << " ";
            }
            cout << endl;
        }
        cout << "~" << endl;
    }
    return 0;
}

